File: Java/16_String-II/E028. Find the Index of the First Occurrence in a String.java
// Find the Index of the First Occurrence in a String --> Z-Function
// https://leetcode.com/problems/find-the-index-of-the-first-occurrence-in-a-string/

class Solution {

    public int strStr(String haystack, String needle) {
        if (needle.isEmpty()) return 0;
        int n = haystack.length(),
            m = needle.length();
        for (int i = 0; i <= n - m; i++) {
            if (haystack.substring(i, i + m).equals(needle)) return i;
        }
        return -1;
    }

    public static void main(String[] args) {
        Solution solution = new Solution();
        String haystack = "hello";
        String needle = "ll";
        int index = solution.strStr(haystack, needle);
        System.out.println("Index of the first occurrence: " + index);
    }
}

File: JavaScript/16_String-II/E028. Find the Index of the First Occurrence in a String.js
// 28. Find the Index of the First Occurrence in a String
// https://leetcode.com/problems/find-the-index-of-the-first-occurrence-in-a-string/

/**
 * Find the first occurrence of needle in haystack
 * @param {string} haystack - String to search in
 * @param {string} needle - String to search for
 * @returns {number} - Index of first occurrence, or -1 if not found
 */
function strStr(haystack, needle) {
    if (needle === "") return 0;

    const n = haystack.length;
    const m = needle.length;

    for (let i = 0; i <= n - m; i++) {
        if (haystack.substring(i, i + m) === needle) {
            return i;
        }
    }

    return -1;
}

// Alternative: Using built-in indexOf
function strStrBuiltin(haystack, needle) {
    return haystack.indexOf(needle);
}

// Test cases
console.log(strStr("hello", "ll")); // 2
console.log(strStr("aaaaa", "bba")); // -1
console.log(strStr("sadbutsad", "sad")); // 0
console.log(strStr("leetcode", "leeto")); // -1
console.log(strStr("", "")); // 0